javascript/fetch-repos.js

```diff
@@ -0,0 +1,142 @@
+    return FourthWall.fetchDefer({
+      url: FourthWall.fileUrl,
+      done: function(result) {
+        var repos = [];
+        if (result.content) {
+          repos = JSON.parse(
+            atob(result.content)
+          ).map(function (item) {
+            // map to ensure gist style keys present
+            // we extend the item to ensure any provided baseUrls are kept
+            return $.extend(item, {
+              'userName': item.owner || item.userName,
+              'repo': item.name ||  item.repo
+            });
+          });
+        }
+        return repos;
+      }
+    });
+  };
+
+  var fetchReposFromGist = function () {
+    return FourthWall.fetchDefer({
+      url: "https://api.github.com/gists/" + FourthWall.gistId,
+      done: function(result) {
+        var repos = [];
+        Object.keys(result.files).forEach(function(file) {
+          var fileData = result.files[file],
+              language = fileData.language;
+          if (file == "users.json") {
+            if (fileData.content) {
+              FourthWall.importantUsers = JSON.parse(fileData.content);
+            }
+          } else if ($.inArray(language, ['JavaScript', 'JSON', null]) !== -1) {
+            repos = JSON.parse(fileData.content);
+          } else if (language === "CSS") {
+            var $custom_css = $('<style>');
+            $custom_css.text( fileData.content );
+            $('head').append( $custom_css );
+          }
+        });
+        return repos;
+      }
+    });
+  };
+
+  var fetchReposFromTeams = function () {
+    var promises = [];
+
+    FourthWall.getTeams().forEach(function(team) {
+      promises.push(fetchReposFromTeam(team));
+    });
+
+    var d = $.Deferred();
+    $.when.apply(null, promises).done(function() {
+      var repos = [].reduce.call(arguments, FetchRepos.mergeRepoArrays, []);
+      d.resolve(repos);
+    });
+
+    return d.promise();
+  };
+
+  var fetchReposFromTeam = function(team) {
+    var d = $.Deferred();
+    fetchTeamId(team).done(function(teamId) {
+      FourthWall.fetchDefer({
+        url: team.baseUrl + "/teams/" + teamId + "/repos",
+        done: function (result) {
+          d.resolve(result.map(function(item) {
+            return {
+              repo: item.name,
+              userName: item.owner.login,
+              baseUrl: team.baseUrl + "/repos",
+            };
+          }));
+        }
+      });
+    });
+    return d;
+  };
+
+  var fetchTeamId = function(team) {
+    return FourthWall.fetchDefer({
+      url: team.baseUrl + '/orgs/' + team.org + '/teams',
+      done: function (result) {
+        for (var i = 0; i < result.length; i++) {
+          if (result[i].name === team.team) {
+            return result[i].id;
+          }
+        }
+      }
+    });
+  };
+
+  // exposed so that it can be tested
+  FourthWall.FetchRepos.mergeRepoArrays = function(repos1, repos2) {
```

Comment:
It'd be a little easier to follow if this method was higher up the file with fetchFromEverywhere